### PR TITLE
Fix incorrect warnings in ParameterList/Dict

### LINKED ISSUE #46983: Constructing a ParameterDict raises a warning
## 🐛 Bug

Constructing a `ParameterDict` raises a `UserWarning`.

## To Reproduce

Install PyTorch 1.7
```py
>>> import torch.nn as nn
>>> nn.ParameterDict()
/Users/fobermey/opt/miniconda3/envs/pyro/lib/python3.7/site-packages/torch/nn/modules/container.py:550: UserWarning: Setting attributes on ParameterDict is not supported.
  warnings.warn("Setting attributes on ParameterDict is not supported.")
ParameterDict()
```

## Expected behavior

PyTorch should not raise warnings in normal usage. Tests should verify this.

## Environment

```
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Mac OSX 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.2)
CMake version: version 3.18.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] numpyro==0.3.0
[pip3] torch==1.7.0
[pip3] torchfile==0.1.0
[pip3] torchvision==0.8.1
[conda] numpy                     1.19.0                   pypi_0    pypi
[conda] numpyro                   0.3.0                    pypi_0    pypi
[conda] torch                     1.7.0                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.8.1                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @albanD @mruberry

### PR DESCRIPTION
**Pull request by CedricPicron**
Fixes #46983.

The solution is based of two components:

1. The introduction of the `_initialized` attribute. This will be used during ParameterList/Dict creation methods `__init__` (introduced in #47772) and  `__setstate__` to not trigger warnings when setting general `Module` attributes. 
2. The introduction of the `not hasattr(self, key)` check to avoid triggering warnings when changing general `Module` attributes such as `.training` during the `train()` and `eval()` methods.

Tests related to the fix are added.

### CODE REVIEW

File: torch/nn/modules/container.py [hunk 1]
```diff
@@ -538,6 +544,11 @@ def __init__(self, parameters: Optional[Mapping[str, 'Parameter']] = None) -> No
         if parameters is not None:
             self.update(parameters)
 
+    def __setstate__(self, state):
+        state['_initialized'] = False
```
Comment (albanD) [inline torch/nn/modules/container.py:548]: nit: I don't think this line is actually needed as we use `getattr` with a default value of `False`.

File: test/test_nn.py [hunk 2]
```diff
@@ -2671,10 +2676,50 @@ def test_parameterlistdict_setting_attributes(self):
...
             torch.nn.utils.weight_norm(mod, "b")
 
+    def test_parameterlistdict_pickle(self):
+        m = nn.ParameterList(map(nn.Parameter, [torch.rand(2), torch.rand(2)]))
+        with warnings.catch_warnings(record=True) as w:
+            m = pickle.loads(pickle.dumps(m))
+        self.assertTrue(len(w) == 0)
+
+        m = nn.ParameterList(map(nn.Parameter, [torch.rand(2), torch.rand(2)]))
+        del m._initialized
```
Comment (albanD) [inline test/test_nn.py:2695]: nit: could you add a comment here to mention that this `del` is here to pretend that this is a Module that was saved with an older version of pytorch?

**Review by albanD — COMMENTED** [review]
Thanks for the PR! It looks good.
Just two small comments and it will be good to go!

**Comment by albanD** [comment]
cc @gchanan This fixes some other cases that I missed in the previous PR.

File: torch/nn/modules/container.py (hunk 1 above)
Comment (CedricPicron) [reply torch/nn/modules/container.py:548]: It is needed whenever a ParameterList/Dict is loaded which had its `_initialized` attribute set to `True`. This line changes it back to `False` as desired. (For those modules not having the `_initialized` attribute, this line is indeed not necessary.)

File: test/test_nn.py (hunk 2 above)
Comment (CedricPicron) [reply test/test_nn.py:2695]: Done.

**Comment by CedricPicron** [comment]
> Thanks for the PR! It looks good.

Thanks for the quick review! I've added a phrase commenting the tests about loading from older checkpoints.

File: torch/nn/modules/container.py (hunk 1 above)
Comment (albanD) [reply torch/nn/modules/container.py:548]: Ho right. Sorry. Looks good.
This whole thing is too complex for what it does... Too bad we don't have a better way to do this warning.

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**